File: solutions/python3_solutions/problems/p1_two_sum.py
from bisect import bisect_left
from typing import Dict, List


def twoSumBrute(nums: List[int], target: int) -> List[int]:
    for i in range(0, len(nums)-1):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    raise Exception('No solution found')

def twoSumSort(nums: List[int], target: int) -> List[int]:
    nums_with_indices = []
    for i in range(len(nums)):
        nums_with_indices.append((nums[i], i))
    nums_with_indices.sort()
    for i in range(len(nums)):
        search_area = nums_with_indices[i+1::]
        desired = target - nums_with_indices[i][0]
        found = bisect_left(search_area, desired, key=lambda a: a[0])
        if found < len(search_area) and search_area[found][0] == desired:
            return [nums_with_indices[i][1], nums_with_indices[found + i + 1][1]]
    raise Exception('No solution found')

def twoSumHash(nums: List[int], target: int) -> List[int]:
    processed_nums: Dict[int, int] = {}
    for i in range(len(nums)):
        desired = target - nums[i]
        if processed_nums.get(desired) is not None:
            return [processed_nums[desired], i]
        processed_nums[nums[i]] = i
    raise Exception('No solution found')

File: solutions/rust-solutions/src/problems/p1_two_sum.rs
use std::collections::HashMap;

pub fn two_sum_brute(nums: Vec<i32>, target: i32) -> Vec<i32> {
    for i in 0..nums.len() {
        for j in i + 1..nums.len() {
            if nums[i] + nums[j] == target {
                return vec![i as i32, j as i32];
            }
        }
    }
    vec![]
}

pub fn two_sum_sort(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut nums_with_indices: Vec<(i32, usize)> = Vec::with_capacity(nums.len());
    for i in 0..nums.len() {
        nums_with_indices.push((nums[i], i as usize));
    }
    nums_with_indices.sort();
    for i in 0..nums_with_indices.len() {
        let desired = target - nums_with_indices[i].0 as i32;
        let found = nums_with_indices[i + 1..].binary_search_by_key(&desired, |&(a, _)| a);
        if found.is_ok() {
            return vec![
                nums_with_indices[i].1 as i32,
                nums_with_indices[found.unwrap() + i + 1].1 as i32,
            ];
        }
    }
    vec![]
}

pub fn two_sum_hash(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut processed_nums = HashMap::new();
    for (i, &v) in nums.iter().enumerate() {
        let desired = target - v;
        if let Some(&j) = processed_nums.get(&desired) {
            return vec![j as i32, i as i32];
        }
        processed_nums.insert(v, i);
    }
    vec![]
}
